File: 백준/Gold/17471. 게리맨더링/게리맨더링.py
# 구상
# DFS로 조합 만들고, UF 로 연결여부 확인 및 최솟값 갱신
# 인원 적으니 인접리스트 대신 인접행렬로

def Find(now):
    if now == parents[now]: return now
    parents[now] = Find(parents[now])
    return parents[now]

def Union(a, b):
    root_A = Find(a)
    root_B = Find(b)
    if root_A == root_B: return

    parents[max(root_A, root_B)] = min(root_A, root_B)

def simulate(team1, team2):
    global min_
    # parents 초기화
    for i in range(N):
        parents[i] = i

    # 구역 합치기
    for i in range(len(team1)):
        for j in range(i+1, len(team1)):
            if adj_matrix[team1[i]][team1[j]] == 1:
                Union(team1[i], team1[j])

    for i in range(len(team2)):
        for j in range(i+1, len(team2)):
            if adj_matrix[team2[i]][team2[j]] == 1:
                Union(team2[i], team2[j])

    # 그룹이 2개 이상이면 break
    group_cnt = 0
    for i in range(N):
        if Find(i) == i:
            group_cnt += 1

    if group_cnt != 2:
        return


    population_1 = 0
    population_2 = 0

    for p in team1:
        population_1 += people[p]

    for p in team2:
        population_2 += people[p]

    min_ = min(min_, abs(population_1 - population_2))




def rec(step, idx, thr):
    if step >= thr:
        mine = []
        opponent = []
        for i in range(N):
            if visit[i] == True:
                mine.append(i)
            else:
                opponent.append(i)
        simulate(mine, opponent)
        return

    for i in range(idx, N):
        visit[i] = True
        rec(step+1, i+1, thr)
        visit[i] = False


N = int(input())

parents = [i for i in range(N)]

people = list(map(int, input().split()))

adj_matrix = [[0]*N for _ in range(N)]
for i in range(N):
    tmp = list(map(lambda x: int(x)-1, input().split()))
    for el in tmp[1:]:
        adj_matrix[i][el] = 1
        adj_matrix[el][i] = 1

min_ = 1001
visit = [False]*N
for num in range(1, (N//2)+1):
    rec(0, 0, num)

if min_ == 1001:
    print(-1)
else:
    print(min_)

File: 백준/Gold/17471. 게리맨더링/게리맨더링.cc
#include <iostream>

#include <queue>

using namespace std;


int N, min_ = 1e8;

int con_info[11][11];

int parent[11], pop_info[11];

int path[11];




int Find(int now) {
	
	if (now == parent[now]) return now;

	return parent[now] = Find(parent[now]);
}

void Union(int a, int b) {
	int root_A = Find(a);
	int root_B = Find(b);

	if (root_A == root_B) return;

	if (root_A > root_B) 
		swap(root_A, root_B);
	
	parent[root_B] = root_A;
}

void discriminate(int n) {

	for (int i = 1; i <= N; i++) {
		parent[i] = i;
	}

	vector<int>n1;
	vector<int>n2;

	for (int i = 1; i <= N; i++) {
		if (path[i] == 1)
			n1.push_back(i);
		else
			n2.push_back(i);
	}


	// n1끼리 연결
	for (int i = 0; i < n1.size(); i++) {
		for (int j = i + 1; j < n1.size(); j++) {
			if (con_info[n1[i]][n1[j]] == 1)
				Union(n1[i], n1[j]);
		}
	}



	// n2끼리 연결
	for (int i = 0; i < n2.size(); i++) {
		for (int j = i + 1; j < n2.size(); j++) {
			if (con_info[n2[i]][n2[j]] == 1)
				Union(n2[i], n2[j]);
		}
	}

	int cnt = 0;

	for (int i = 1; i <= N; i++) {
		if (parent[i] == i) cnt++;
	}

	if (cnt != 2) return;

	int pop1 = 0;
	int pop2 = 0;

	for (int i = 0; i < n1.size(); i++) {
		pop1 += pop_info[n1[i]];
	}

	for (int i = 0; i < n2.size(); i++) {
		pop2 += pop_info[n2[i]];
	}

	min_ = min(min_, abs(pop1 - pop2));
}



void DFS(int n, int now, int cnt) {

	if (cnt == n) {
		// 이어지는 지 판별
		discriminate(n);
	}

	for (int i = now; i <= N; i++) {

		if (path[i] == 1) continue;
		
		path[i] = 1;
		DFS(n, i + 1, cnt + 1);
		path[i] = 0;
	}


}

void solve() {

	for (int i = 1; i <= N / 2; i++) {
		DFS(i,1,0);
	}


}




int main(void)
{
	cin >> N;

	for (int i = 1; i <= N; i++) {
		cin >> pop_info[i];
	}

	for (int i = 1; i <= N; i++) {
		int num;
		cin >> num;
		for (int j = 0; j < num; j++) {
			int to;
			cin >> to;
			con_info[i][to] = 1;
		}
	}

	solve();

	if (min_ == 1e8)
		cout << -1;
	else
		cout << min_;





}